### PR TITLE
convert @babel/types to typescript

### PR DESCRIPTION
**Pull request by zxbodya**
| Q                        | A 
| ------------------------ | ---
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12431"></a>

@babel/types part of #11578 

---

EDIT by @nicolo-ribaudo 

This should be merged as two different commits, to help git track file renames (useful for things like `git blame`):
- fbf88ccb52f184111e12b26d66169bae7a06e795
- Everything else

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -51,8 +51,7 @@ build-flow-typings:
 	$(NODE) packages/babel-types/scripts/generators/flow.js > packages/babel-types/lib/index.js.flow
 
 build-typescript-typings:
-	$(NODE) packages/babel-types/scripts/generators/typescript.js > packages/babel-types/lib/index.d.ts
-	$(NODE) packages/babel-types/scripts/generators/typescript.js --ts3.7 > packages/babel-types/lib/index-ts3.7.d.ts
+	echo "TODO: maybe add types for typescript 3.7"
```
Comment (JLHwung) [inline Makefile:54]: We can add the `index-ts3.7.d.ts` to `/src` and keep it for backward compatibilities. Just adding a note that this file is no longer updated to reflect most recent changes in `@babel/types` -- The interface definition may be incomplete but it should be good if you don't use the new features of `@babel/types`.

Folks can pin Babel types for TypeScript 3.7 support or update babel types and typescript to 4.0 as well.

File: Makefile [hunk 2]
```diff
@@ -149,6 +148,10 @@ clean-tsconfig:
 	rm -f packages/*/tsconfig.{tsbuildinfo,json}
 	rm -f codemods/*/tsconfig.{tsbuildinfo,json}
 	rm -f eslint/*/tsconfig.{tsbuildinfo,json}
+	# revert files in git
```
Comment (JLHwung) [inline Makefile:151]: I don't think we have added any `tsconfig.json` to git. Why we have to revert here? 

@zxbodya If it is for rebasing the `ts` branch, you can split it in a separate commit used by that branch only.

Comment (zxbodya) [reply Makefile]: there is separate PR #12420  - this is needed when converting other packages, for tsconfigs that can not be generated

File: packages/babel-types/src/builders/builder.ts [hunk 3]
```diff
@@ -1,9 +1,12 @@
-// @flow
 import loClone from "lodash/clone";
 import { NODE_FIELDS, BUILDER_KEYS } from "../definitions";
 import validate from "../validators/validate";
+import type * as types from "../types";
```
[the remark this answers is not in the document]
Comment (zxbodya) [reply packages/babel-types/src/builders/builder.ts]: updated it to use `t` instead

File: Makefile (hunk 1 above)
Comment (zxbodya) [reply Makefile:54]: other option would be to bundle generated type definitions into single file using something like [rollup-plugin-dts](https://github.com/Swatinem/rollup-plugin-dts), and then to transpile it to older typescript using  [downlevel-dts](https://github.com/sandersn/downlevel-dts)

but - I think, this is not something needed - if updating babel? why not update typescript? (especially considering that types are more complete using it)

typically typescript is very easy to update, and honestly - I do not really imagine there is a good case when babel is to be updated, while keeping old typescript.

my preference would be - to support only the last typescript version, and to not complicate things

**Comment by nicolo-ribaudo** [comment]
@zxbodya I'll open a PR to your fork later today to fix the e2e test failure.

File: Makefile (hunk 1 above)
Comment (nicolo-ribaudo) [reply Makefile:54]: > I do not really imagine there is a good case when babel is to be updated, while keeping old typescript.

Usually Babel is installed with `^` in `package.json`, while different projects install TS with `~` because it doesn't follow semver.

If you had this `package.json` from last year:
```js
"@babel/types": "^7.6.0",
"typescript": "~3.7.0`
```
it would now be broken :shrug:

I agree with @JLHwung: we can stop generating `index-ts3.7.d.ts` (we can just hard-code it and stop updating it), but we cannot remove it in a minor version.

File: packages/babel-types/src/builders/generated/deprecated.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
+/*
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/builders/generated/deprecated.js:1]: Why don't generate this as `.ts`?

Also, having `deprecated.d.ts` in `src` generates a `deprecated.d.js` file in `lib`.

File: packages/babel-types/lib/builders/generated/deprecated.d.ts [hunk 5]
```diff
@@ -0,0 +1,2 @@
+// exports with old naming are not expected to be used from typescript
```
Comment (nicolo-ribaudo) [inline packages/babel-types/lib/builders/generated/deprecated.d.ts:1]: Why is this file committed?

File: packages/babel-types/src/builders/generated/deprecated.js (hunk 4 above)
Comment (zxbodya) [reply packages/babel-types/src/builders/generated/deprecated.js:1]: names of exports here are conflicting with names of AST nodes, and so to export only ast nodes in type definitions deprecated builders (starting with upper case, were placed in js file, to not include then in typechecks)

File: packages/babel-types/lib/builders/generated/deprecated.d.ts (hunk 5 above)
Comment (zxbodya) [reply packages/babel-types/lib/builders/generated/deprecated.d.ts:1]: there is `src/builders/generated/deprecated.d.ts` but typescript does not output `d.ts` files from other `d.ts` files, this a workaround to add the file into lib folder

File: packages/babel-types/src/builders/generated/deprecated.js [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+ * This file is auto-generated! Do not modify it directly.
+ * To re-generate run 'make build'
+ */
+
+export { arrayExpression as ArrayExpression } from "./index";
+export { assignmentExpression as AssignmentExpression } from "./index";
+export { binaryExpression as BinaryExpression } from "./index";
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/builders/generated/deprecated.js:8]: To save some space, this could be
```js
export {
  arrayExpression as ArrayExpression,
  assignmentExpression as AssignmentExpression,
  ...
} from "./index";
```

File: packages/babel-types/src/modifications/appendToMemberExpression.ts [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
 
 /**
  * Append a node to a member expression.
  */
-export default function appendToMemberExpression<T: Object>(
+export default function appendToMemberExpression<T extends t.MemberExpression>(
   member: T,
-  append: Object,
-  computed?: boolean = false,
+  append: t.MemberExpression["property"],
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/modifications/appendToMemberExpression.ts:9]: I wonder if this should just be

```ts
export default function appendToMemberExpression(
  member: t.MemberExpression,
  append: t.MemberExpression["property"],
  computed: boolean = false,
```

File: packages/babel-types/src/modifications/flow/removeTypeDuplicates.ts [hunk 8]
```diff
@@ -1,17 +1,24 @@
...
   isFlowBaseAnnotation,
+  isIdentifier,
 } from "../../validators/generated";
+import type * as t from "../../types";
+
+function getQualificationName(node: t.GenericTypeAnnotation["id"]) {
+  return isIdentifier(node)
+    ? node.name
+    : `${node.id.name}.${getQualificationName(node.qualification)}`;
+}
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/modifications/flow/removeTypeDuplicates.ts:14]: Uh is this a bugfix caught by TS?

Btw, nit but the function name should be `getQualifiedName`.

File: packages/babel-types/src/builders/generated/deprecated.js (hunk 6 above)
Comment (zxbodya) [reply packages/babel-types/src/builders/generated/deprecated.js:8]: will do

File: packages/babel-types/src/modifications/prependToMemberExpression.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import { memberExpression } from "../builders/generated";
+import type * as t from "../types";
+
+/**
+ * Prepend a node to a member expression.
+ */
+export default function prependToMemberExpression<
+  T extends Pick<t.MemberExpression, "object" | "property">
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/modifications/prependToMemberExpression.ts:8]:
```js
Pick<t.MemberExpression, "object" | "property">
```

I love this kind of types, but isn't this the same as `t.Expression`?

File: packages/babel-types/src/modifications/prependToMemberExpression.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+>(member: T, prepend: t.MemberExpression["object"]): T {
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/modifications/prependToMemberExpression.ts:9]: Also `t.MemberExpression["object"]` is probably `t.Expression`

File: packages/babel-types/src/modifications/removeProperties.ts [hunk 11]
```diff
@@ -12,8 +12,8 @@ const CLEAR_KEYS_PLUS_COMMENTS = COMMENT_KEYS.concat(["comments"]).concat(
  * properties like location data and raw token data.
  */
 export default function removeProperties(
-  node: Object,
-  opts?: Object = {},
+  node: t.Node,
+  opts: { preserveComments?: boolean } = {},
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/modifications/removeProperties.ts:16]: :heart:

File: packages/babel-types/src/modifications/typescript/removeTypeDuplicates.ts [hunk 12]
```diff
@@ -7,9 +7,7 @@ import {
 /**
  * Dedupe type annotations.
  */
-export default function removeTypeDuplicates(
-  nodes: Array<Object>,
-): Array<Object> {
+export default function removeTypeDuplicates(nodes: Array<any>): Array<any> {
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/modifications/typescript/removeTypeDuplicates.ts:10]:
```js
function removeTypeDuplicates(nodes: Array<t.Node>): Array<t.Node>
```

File: packages/babel-types/src/retrievers/getBindingIdentifiers.ts [hunk 13]
```diff
@@ -1,20 +1,39 @@
...
+  node: t.Node,
+  duplicates?: false,
+  outerOnly?: boolean,
+): Record<string, t.Identifier>;
+
+function getBindingIdentifiers(
+  node: t.Node,
+  duplicates?: boolean,
+  outerOnly?: boolean,
+): Record<string, t.Identifier | Array<t.Identifier>>;
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/retrievers/getBindingIdentifiers.ts:27]: Q: Do we need this, or is it implied by the two above?

Also, if we need it the return type should probably be `Record<string, t.Identifier> | Record<string, Array<t.Identifier>>`.

File: packages/babel-types/src/retrievers/getBindingIdentifiers.ts [hunk 14]
```diff
@@ -119,3 +138,5 @@ getBindingIdentifiers.keys = {
   VariableDeclaration: ["declarations"],
   VariableDeclarator: ["id"],
 };
+
+export default getBindingIdentifiers;
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/retrievers/getBindingIdentifiers.ts:142]: Q: Why was `export default` moved?

File: packages/babel-types/src/retrievers/getOuterBindingIdentifiers.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+export default function getOuterBindingIdentifiers(
+  node: t.Node,
+  duplicates?: false,
+): Record<string, t.Identifier>;
+
+export default function getOuterBindingIdentifiers(
+  node: t.Node,
+  duplicates: boolean,
+): Record<string, t.Identifier | Array<t.Identifier>>;
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/retrievers/getOuterBindingIdentifiers.ts:17]: Same comment as for `packages/babel-types/src/retrievers/getBindingIdentifiers.ts`.

File: packages/babel-types/src/traverse/traverse.ts [hunk 16]
```diff
@@ -1,15 +1,22 @@
...
+  key: string;
+  index?: number;
 }>;
-export type TraversalHandler<T> = (BabelNode, TraversalAncestors, T) => void;
+
+export type TraversalHandler<T> = (
+  this: undefined,
+  node: t.Node,
+  parent: TraversalAncestors,
+  type: T,
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/traverse/traverse.ts:14]:
```suggestion
  state: T,
```

File: packages/babel-types/src/traverse/traverseFast.ts [hunk 17]
```diff
@@ -1,14 +1,14 @@
...
  * and processing.
  */
 export default function traverseFast(
-  node: Object,
-  enter: (node: BabelNode, opts?: Object) => void,
-  opts?: Object,
+  node: t.Node | null | undefined,
+  enter: (node: t.Node, opts?: any) => void,
+  opts?: any,
 ): void {
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/traverse/traverseFast.ts:12]: We could parametrize `opts` to `T` rather than `any`, so that the type is "forwarded" to the callback.

File: packages/babel-types/src/types.ts [hunk 18]
```diff
@@ -0,0 +1,2518 @@
+// NOTE: This file is autogenerated. Do not modify.
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/types.ts:1]: Could we move this file to `packages/babel-types/src/types/generated/index.ts`, similarly to how we do it for other generated files? :pray:

File: packages/babel-types/src/utils/inherit.ts [hunk 19]
```diff
@@ -1,9 +1,4 @@
-// @flow
-export default function inherit(
-  key: string,
-  child: Object,
-  parent: Object,
-): void {
+export default function inherit(key: string, child: any, parent: any): void {
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/utils/inherit.ts:1]: `t.Node` instead of `any`.

File: packages/babel-types/src/utils/react/cleanJSXElementLiteralChild.ts [hunk 20]
```diff
@@ -1,9 +1,10 @@
-// @flow
 import { stringLiteral } from "../../builders/generated";
 
 export default function cleanJSXElementLiteralChild(
-  child: { value: string },
-  args: Array<Object>,
+  child: {
+    value: string;
+  },
+  args: Array<any>,
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/utils/react/cleanJSXElementLiteralChild.ts:7]: `t.Node`

File: packages/babel-types/src/validators/isPlaceholderType.ts [hunk 21]
```diff
@@ -10,7 +9,8 @@ export default function isPlaceholderType(
 ): boolean {
   if (placeholderType === targetType) return true;
 
-  const aliases: ?Array<string> = PLACEHOLDERS_ALIAS[placeholderType];
+  const aliases: Array<string> | undefined | null =
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/validators/isPlaceholderType.ts:12]: This is never `null`

**Comment by nicolo-ribaudo** [comment]
(I'm preparing a longer review)

File: packages/babel-types/src/validators/is.ts [hunk 22]
```diff
@@ -1,15 +1,41 @@
...
 import isType from "./isType";
 import isPlaceholderType from "./isPlaceholderType";
 import { FLIPPED_ALIAS_KEYS } from "../definitions";
+import type * as t from "../types";
 
+export default function is<T extends t.Node["type"]>(
+  type: T,
+  node: t.Node | null | undefined,
+  opts?: undefined,
+): node is Extract<t.Node, { type: T }>;
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/validators/is.ts:11]: Neat!

File: packages/babel-types/src/validators/isType.ts [hunk 23]
```diff
@@ -1,17 +1,31 @@
...
+  targetType: string,
+): boolean {
   if (nodeType === targetType) return true;
 
   // This is a fast-path. If the test above failed, but an alias key is found, then the
   // targetType was a primary node type, so there's no need to check the aliases.
   if (ALIAS_KEYS[targetType]) return false;
 
-  const aliases: ?Array<string> = FLIPPED_ALIAS_KEYS[targetType];
+  const aliases: Array<string> | undefined | null =
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/validators/isType.ts:27]: This is never `null`

**Review by nicolo-ribaudo — COMMENTED** [review]
Overall, this is awesome!

File: packages/babel-types/src/builders/typescript/createTSUnionType.ts [hunk 24]
```diff
@@ -5,9 +5,7 @@ import removeTypeDuplicates from "../../modifications/typescript/removeTypeDupli
  * Takes an array of `types` and flattens them, removing duplicates and
  * returns a `UnionTypeAnnotation` node containing them.
  */
-export default function createTSUnionType(
-  typeAnnotations: Array<Object>,
-): Object {
+export default function createTSUnionType(typeAnnotations: Array<any>): any {
```
[the remark this answers is not in the document]
Comment (zxbodya) [reply packages/babel-types/src/builders/typescript/createTSUnionType.ts:8]: interesting one - when added types, found few bugs - will add fix to pr 🙈

File: packages/babel-types/src/converters/toBlock.ts [hunk 25]
```diff
@@ -10,8 +9,12 @@ import {
   expressionStatement,
   blockStatement,
 } from "../builders/generated";
+import type * as t from "../types";
 
-export default function toBlock(node: Object, parent: Object): Object {
+export default function toBlock(
+  node: t.Statement | t.Expression,
+  parent?: t.Node | null,
```
[the remark this answers is not in the document]
Comment (zxbodya) [reply packages/babel-types/src/converters/toBlock.ts:16]: not sure… I suspect somewhere it was called with null and just added it to match the usage

but it looks to be not used anymore - removing the null

File: packages/babel-types/src/converters/toComputedKey.ts [hunk 26]
```diff
@@ -1,11 +1,11 @@
-// @flow
 import { isIdentifier } from "../validators/generated";
 import { stringLiteral } from "../builders/generated";
+import type * as t from "../types";
 
 export default function toComputedKey(
-  node: Object,
-  key: Object = node.key || node.property,
-): Object {
+  node: any,
```
[the remark this answers is not in the document]
Comment (zxbodya) [reply packages/babel-types/src/converters/toComputedKey.ts:6]: adding todo to come back to this later

File: packages/babel-types/src/converters/toExpression.ts [hunk 27]
```diff
@@ -1,12 +1,20 @@
...
-export default function toExpression(node: Object): Object {
+export default function toExpression(node: t.Function): t.FunctionExpression;
+export default function toExpression(node: t.Class): t.ClassExpression;
+export default function toExpression(
+  node: t.ExpressionStatement | t.Expression | t.Class | t.Function,
+): t.Expression;
+
+export default function toExpression(
+  node: t.ExpressionStatement | t.Expression | t.Class | t.Function,
+): t.Expression {
```
[the remark this answers is not in the document]
Comment (zxbodya) [reply packages/babel-types/src/converters/toExpression.ts:17]: the reason for this is that when using function overloads, the implementation is not exported as function type, that is why it duplicated

will refactor a bit to make this less duplicated

File: packages/babel-types/src/converters/toStatement.ts [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+  isClass,
+  isAssignmentExpression,
+} from "../validators/generated";
+import { expressionStatement } from "../builders/generated";
+import type * as t from "../types";
+
+export default function toStatement(
+  node: t.AssignmentExpression,
+  ignore?: boolean,
+): t.ExpressionStatement;
```
[the remark this answers is not in the document]
Comment (zxbodya) [reply packages/babel-types/src/converters/toStatement.ts:13]: got all of the overloads for `toStatement ` here from https://github.com/babel/babel/blob/main/packages/babel-types/scripts/generators/typescript.js#L213

have not really looked into improving it

File: packages/babel-types/src/converters/valueToNode.ts [hunk 29]
```diff
@@ -15,8 +14,33 @@ import {
   unaryExpression,
   binaryExpression,
 } from "../builders/generated";
+import type * as t from "../types";
 
-export default function valueToNode(value: any): Object {
+export default function valueToNode(value: undefined): t.Identifier; // (should this not be a types.UnaryExpression to avoid shadowing?)
```
[the remark this answers is not in the document]
Comment (zxbodya) [reply packages/babel-types/src/converters/valueToNode.ts:19]: adding todo
btw, the comment came from here: https://github.com/babel/babel/blob/main/packages/babel-types/scripts/generators/typescript.js#L224

File: packages/babel-types/src/converters/valueToNode.ts [hunk 30]
```diff
@@ -15,8 +14,33 @@ import {
...
+export default function valueToNode(value: null): t.NullLiteral;
+export default function valueToNode(value: string): t.StringLiteral;
+// Infinities and NaN need to use a types.BinaryExpression; negative values must be wrapped in types.UnaryExpression
+export default function valueToNode(
+  value: number,
+): t.NumericLiteral | t.BinaryExpression | t.UnaryExpression;
+export default function valueToNode(value: RegExp): t.RegExpLiteral;
+export default function valueToNode(
+  value: ReadonlyArray<
+    undefined | boolean | null | string | number | RegExp | object
```
[the remark this answers is not in the document]
Comment (zxbodya) [reply packages/babel-types/src/converters/valueToNode.ts:30]: got it from here: https://github.com/babel/babel/blob/main/packages/babel-types/scripts/generators/typescript.js#L232

it indeed looks to be like `any` 🤔

File: packages/babel-types/src/definitions/utils.ts [hunk 31]
```diff
@@ -21,12 +32,16 @@ function getType(val) {
 }
 
 // TODO: Import and use Node instead of any
-type Validator = (any, string, any) => void;
+type Validator = { chainOf?: Validator[] } & ((
+  c: any,
+  b: string,
+  a: any,
```
[the remark this answers is not in the document]
Comment (zxbodya) [reply packages/babel-types/src/definitions/utils.ts:38]: changing argument names, keeping it as any for now because making it `t.Node` causes TS errors in usages where more granular type is needed

File: packages/babel-types/src/modifications/appendToMemberExpression.ts (hunk 7 above)
Comment (zxbodya) [reply packages/babel-types/src/modifications/appendToMemberExpression.ts:9]: yeah, looks reasonable - updating it…

not sure why in flow sources it was made as generic

File: packages/babel-types/src/modifications/flow/removeTypeDuplicates.ts (hunk 8 above)
Comment (zxbodya) [reply packages/babel-types/src/modifications/flow/removeTypeDuplicates.ts:14]: yes, the first one I actually found with TS :)
updating the name

File: packages/babel-types/src/traverse/traverseFast.ts (hunk 17 above)
Comment (zxbodya) [reply packages/babel-types/src/traverse/traverseFast.ts]: added todo - it does not really align with default value for it being empty object

File: packages/babel-types/src/utils/inherit.ts (hunk 19 above)
Comment (zxbodya) [reply packages/babel-types/src/utils/inherit.ts]: done

File: packages/babel-types/src/utils/react/cleanJSXElementLiteralChild.ts (hunk 20 above)
Comment (zxbodya) [reply packages/babel-types/src/utils/react/cleanJSXElementLiteralChild.ts]: done

File: packages/babel-types/src/validators/isType.ts (hunk 23 above)
Comment (zxbodya) [reply packages/babel-types/src/validators/isType.ts]: updated
this btw, is coming from from converted flow type: `?Array<string>` ==> `null | undefined | Array<string>`

File: packages/babel-types/src/builders/typescript/createTSUnionType.ts [hunk 32]
```diff
@@ -1,14 +1,15 @@
...
  * Takes an array of `types` and flattens them, removing duplicates and
  * returns a `UnionTypeAnnotation` node containing them.
  */
 export default function createTSUnionType(
-  typeAnnotations: Array<Object>,
-): Object {
-  const types = typeAnnotations.map(type => type.typeAnnotations);
+  typeAnnotations: Array<t.TSTypeAnnotation>,
+): t.TSType {
+  const types = typeAnnotations.map(type => type.typeAnnotation);
```
Comment (zxbodya) [inline packages/babel-types/src/builders/typescript/createTSUnionType.ts:12]: looking at usage from babel-traverse apparently this was a typo and instead it should be `. typeAnnotation` instead

File: Makefile [hunk 33]
```diff
@@ -45,13 +45,11 @@ generate-standalone:
...
 
-build-typings: build-flow-typings build-typescript-typings
-
 build-flow-typings:
 	$(NODE) packages/babel-types/scripts/generators/flow.js > packages/babel-types/lib/index.js.flow
 
-build-typescript-typings:
-	echo "TODO: maybe add types for typescript 3.7"
+build-typescript-3.7-typings:
+	$(NODE) packages/babel-types/scripts/generators/typescript-3.7.js > packages/babel-types/lib/index-ts3.7.d.ts
```
Comment (zxbodya) [inline Makefile:52]: would like to not have duplicated script for typescript type generation… - going to try https://github.com/sandersn/downlevel-dts

**Comment by nicolo-ribaudo** [comment]
Here you can see the published package - the published types look good.

[types-7.12.10.tgz](https://github.com/babel/babel/files/5666897/types-7.12.10.zip)

File: Makefile (hunk 33 above)
Comment (nicolo-ribaudo) [reply Makefile:52]: I'd prefer to keep this solution that has been working well for a while. We'll remove it in Babel 8.

File: packages/babel-types/package.json [hunk 34]
```diff
@@ -14,6 +14,13 @@
     "directory": "packages/babel-types"
   },
   "main": "lib/index.js",
+  "typesVersions": {
+    ">=3.7": {
```
Comment (zxbodya) [inline packages/babel-types/package.json:18]: should this be `<=3.7`? (to not override generated types for new TS)

Comment (nicolo-ribaudo) [reply packages/babel-types/package.json:18]: Oh good catch!

Comment (nicolo-ribaudo) [reply packages/babel-types/package.json:18]: Oh wait, this means that typings for 3.7 have never worked :thinking: 

@JLHwung Given this, maybe it's fine to just forget about them?

Comment (nicolo-ribaudo) [reply packages/babel-types/package.json:18]:
```suggestion
    "<3.7": {
```

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
I'm rebasing

File: Gulpfile.js [hunk 35]
```diff
@@ -102,6 +113,8 @@ function generateTypeHelpers(helperKind) {
       })
     )
     .pipe(gulp.dest(dest));
+
+  return finish(stream);
```
Comment (nicolo-ribaudo) [inline Gulpfile.js:117]: @JLHwung This is a bug introduced by the recent gulpfile changes: we were returning a stream instead of a promise, and we were passing those streams to `Promise.all`.

Comment (JLHwung) [reply Gulpfile.js:117]: Ah good catch!

**Review by JLHwung — APPROVED** [review]
This is awesome!

**Comment by nicolo-ribaudo** [comment]
:shipit:
